crates/sui-indexer-alt/migrations/2024-10-16-225607_watermarks/up.sql

```diff
@@ -0,0 +1,28 @@
+CREATE TABLE IF NOT EXISTS watermarks
+(
+    -- The pipeline governed by this watermark, i.e `epochs`, `checkpoints`, `transactions`.
+    pipeline                    TEXT          PRIMARY KEY,
+    -- Inclusive upper epoch bound for this entity's data. Committer updates this field. Pruner uses
+    -- this to determine if pruning is necessary based on the retention policy.
+    epoch_hi_inclusive          BIGINT        NOT NULL,
+    -- Inclusive upper checkpoint bound for this entity's data. Committer updates this field. All
+    -- data of this entity in the checkpoint must be persisted before advancing this watermark. The
+    -- committer refers to this on disaster recovery to resume writing.
+    checkpoint_hi_inclusive     BIGINT        NOT NULL,
+    -- Exclusive upper transaction sequence number bound for this entity's data. Committer updates
+    -- this field.
+    tx_hi                       BIGINT        NOT NULL,
+    -- Inclusive lower epoch bound for this entity's data. Pruner updates this field when the epoch range exceeds the retention policy.
+    epoch_lo                    BIGINT        NOT NULL,
+    -- Inclusive low watermark that the pruner advances. Corresponds to the epoch id, checkpoint
+    -- sequence number, or tx sequence number depending on the entity. Data before this watermark is
+    -- considered pruned by a reader. The underlying data may still exist in the db instance.
+    reader_lo                   BIGINT        NOT NULL,
+    -- Updated using the database's current timestamp when the pruner sees that some data needs to
+    -- be dropped. The pruner uses this column to determine whether to prune or wait long enough
+    -- that all in-flight reads complete or timeout before it acts on an updated watermark.
+    timestamp_ms                BIGINT        NOT NULL,
+    -- Column used by the pruner to track its true progress. Data at and below this watermark can
+    -- be immediately pruned.
+    pruner_hi                   BIGINT        NOT NULL
```

Comment:
btw, should this be `pruner_hi_inclusive` instead?

Reply:
No, it's the exclusive upperbound from the pruner's perspective -- this is actually `pruner_lo` from your PR, but I renamed it based on our discussion on Wednesday because it's actually the (exclusive) upperbound for the pruner to meet.

Reply:
Ah alright, was going off the comments

Should I standardize on the current indexer to pruner_hi as well? I renamed it to pruner_hi_inclusive

Reply:
Ah good call, I think it's the comment that should be updated in that case (and in that case the name in the OG Indexer as well) -- if the algorithm is that this value gets set from `reader_lo` then it must be an exclusive upperbound for the indexer, otherwise we would be off by one on our pruning.